### PR TITLE
Gradle multi module, changes to watch child project

### PR DESCRIPTION
**Pull request by arunvenmany-ibm**
In gradle multi modules, changes added to
1. Watch for changes in webapp source directory for 'war' child project
2. Watch for changes in java for all source and test source directories for all child projects
3. Watch for changes in build.gradle for all child projects

### CODE REVIEW

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy [hunk 1]
```diff
@@ -1215,6 +1216,7 @@ class DevTask extends AbstractFeatureTask {
 
     @TaskAction
     void action() {
+        HashSet<Project> allProjects = new HashSet<>()
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1219]: Shouldn't the variable type just be `Set<Project>`? And why do you not include `Project` when you create the `HashSet`?

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy [hunk 2]
```diff
@@ -1262,15 +1264,19 @@ class DevTask extends AbstractFeatureTask {
 
         File buildFile = project.getBuildFile();
 
+        List<Path> webResourceDirs = LooseWarApplication.getWebSourceDirectoriesToMonitor(project);
         // Instantiate util before any child gradle tasks launched so it can help find available port if needed
+        List<ProjectModule> projectModules = getProjectModules()
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1269]: Can you add some comments here to explain the difference between projectModules and allProjects? And what they are going to be used for?

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy [hunk 3]
```diff
@@ -1410,6 +1416,84 @@ class DevTask extends AbstractFeatureTask {
         }
     }
 
+    private List<ProjectModule> getProjectModules() {
+        List<ProjectModule> upstreamProjects = new ArrayList<ProjectModule>();
+
+        HashSet<Project> allProjects = new HashSet<>()
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1422]: same comment about `Set<Project>` and creating `HashSet` with `<Project>`.

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy [hunk 4]
```diff
@@ -1262,15 +1264,19 @@ class DevTask extends AbstractFeatureTask {
 
         File buildFile = project.getBuildFile();
 
+        List<Path> webResourceDirs = LooseWarApplication.getWebSourceDirectoriesToMonitor(project);
         // Instantiate util before any child gradle tasks launched so it can help find available port if needed
+        List<ProjectModule> projectModules = getProjectModules()
+        allProjects.add(project)
+        allProjects.addAll(DevTaskHelper.getAllUpstreamProjects(project))
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1271]: The call to `getProjectModules` seems to be doing this exact same thing...why? Can we pass in a collection to that method and let it populate `allProjects` for us?

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy [hunk 5]
```diff
@@ -1410,6 +1416,84 @@ class DevTask extends AbstractFeatureTask {
...
+            JavaCompilerOptions upstreamCompilerOptions = new JavaCompilerOptions();
+            SourceSet mainSourceSet = dependencyProject.sourceSets.main;
+            SourceSet testSourceSet = dependencyProject.sourceSets.test;
+            DefaultFilePropertyFactory.DefaultDirectoryVar outputDirectory = mainSourceSet.java.classesDirectory;
+            DefaultFilePropertyFactory.DefaultDirectoryVar testOutputDirectory = testSourceSet.java.classesDirectory;
+            Set<String> compileArtifacts = new HashSet<String>();
+            Set<String> testArtifacts = new HashSet<String>();
+            File upstreamSourceDir = new File(mainSourceSet.java.sourceDirectories.asPath)
+            File upstreamOutputDir = outputDirectory.asFile.get();
+            File upstreamTestSourceDir = new File(testSourceSet.java.sourceDirectories.asPath);
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1436]: You are referencing a lot of different configurations in the preceding lines....could some of them not exist and end up throwing an exception?

File: src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            throw new GradleException("Archive path not found. Supported formats are jar, war, ear, and springboot jar.")
+        }
+    }
+
+    /**
+     * get deploy dependencies for gradle project
+     * @param project project
+     * @return
+     */
+    public static HashMap<File, Dependency> getDeployDependencies(Project project) {
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy:61]: Should return a `Map` not a `HashMap`.

File: src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        HashMap<File, Dependency> completeDeployDeps = new HashMap<File, Dependency>();
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy:62]: variable should be a `Map`

File: src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        File[] filesAsDeps = project.configurations.deploy.getFiles().toArray()
+        Dependency[] deployDeps = project.configurations.deploy.getAllDependencies().toArray()
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy:64]: Do we need to check for `deploy` configuration exists before accessing methods on it?

File: src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+        if (filesAsDeps.size() == deployDeps.size()) {
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy:66]: Can `filesAsDeps` be null?

Comment (cherylking) [reply src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy:66]: And why do we need `filesAsDeps` and `deployDeps` size to be the same? If they are not, then we return no dependencise?

File: src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy [hunk 10]
```diff
@@ -32,4 +38,56 @@ public class LooseWarApplication extends LooseApplication {
         config.addDir(sourceDir, "/")
     }
 
+    /**
+     * get web app source directories
+     * @param project
+     * @return
+     */
+    public static List<Path> getWebSourceDirectoriesToMonitor(Project project) {
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy:46]: Do these static methods belong here? Seems like they belong on a helper/utility class.

File: src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarAppDevMode.groovy [hunk 11]
```diff
@@ -0,0 +1,65 @@
+/*
+ * (C) Copyright IBM Corporation 2020, 2023.
```
Comment (cherylking) [inline src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarAppDevMode.groovy:2]: change copyright to just say 2024 or remove it entirely (I personally don't think test cases need a copyright).

**Review by cherylking — CHANGES_REQUESTED** [review]

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy (hunk 2 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1269]: ProjectModules are constructed passed to DevUtils for multi module inside DevUtils. its used in multiple areas for watching sub project files

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy (hunk 4 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1271]: allProjects contains projectmodules plus the ear project

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy (hunk 5 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1436]: All these will be always present for any java project. We are using similarly in LooseEarApplication

File: src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy (hunk 6 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy:61]: Done

File: src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy (hunk 8 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/utils/DevTaskHelper.groovy:64]: We are doing the check in the calling function, LooseWarApplication#52

File: src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarAppDevMode.groovy (hunk 11 above)
Comment (arunvenmany-ibm) [reply src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarAppDevMode.groovy]: Removed Copyright

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy (hunk 1 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy]: Changed to Set

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy (hunk 3 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy]: Corrected

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy (hunk 4 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1276]: I have removed those. Please check now..
We need list of all dependent projects to construct the projectmodule object
We need list of all dependent project+ current ear project in allProjects variable, which is not used now, its going to be used in updateArtifacts in further PRs

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy (hunk 2 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy:1274]: Maybe I can remove allProjects now, and add it later to avoid confusion on where its used

File: src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+        expression = "/archive/archive";
+        nodes = (NodeList) xPath.compile(expression).evaluate(inputDoc, XPathConstants.NODESET);
+        Assert.assertEquals("Number of <archive/> element ==>", 2, nodes.getLength());
+        if (OSUtil.isWindows()) {
+            ejbWar = "//ejb-war-1.0-SNAPSHOT.war"
+            warWebappsFolder = "//multi-module-loose-ear-pages-test//war//src//main//webapp"
+            ejbJar = "//WEB-INF//lib//ejb-jar-1.0-SNAPSHOT.jar"
+        }
+        Assert.assertTrue(List.of(nodes.item(0).getAttributes()
```
Comment (cherylking) [inline src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy:97]: This is failing in the build:

Screenshot 2024-10-16 at 10 51 23 AM

**Review by cherylking — CHANGES_REQUESTED** [review]

File: src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy (hunk 12 above)
Comment (cherylking) [reply src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy:97]: probably not available with Java 8

File: src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy [hunk 13]
```diff
@@ -388,7 +387,6 @@ class DevTask extends AbstractFeatureTask {
                     null /* compileOptions not needed since useBuildRecompile is true */, keepTempContainerfile, mavenCacheLocation, projectModuleList /* multi module upstream projects */,
                     false /* recompileDependencies only supported in ci.maven */, packagingType, buildFile, null /* parent build files */, generateFeatures, null /* compileArtifactPaths */, null /* testArtifactPaths */, webResourceDirs /* webResources */
                 );
-            this.upstreamGradleProjects = gradleProjects;
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy]: I don't understand why you made all these changes unrelated to windows and the yaml. Only the test case needed to be fixed for Windows and Java 8.

Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy]: Removed the allProjects from this PR. Will add this later when its being used

File: src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy [hunk 14]
```diff
@@ -2,10 +2,16 @@ package io.openliberty.tools.gradle.utils;
...
+import org.apache.commons.io.FilenameUtils
+import org.gradle.api.Project
 import org.gradle.api.Task
+import org.gradle.api.artifacts.Dependency
+import org.gradle.api.artifacts.ProjectDependency
 import org.gradle.api.logging.Logger
 import org.gradle.api.tasks.bundling.War
 
+import java.nio.file.Path
+
```
Comment (cherylking) [inline src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy:14]: You should undo the changes to this file.

File: src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+            warWebappsFolder = "\\multi-module-loose-ear-pages-test\\war\\src\\main\\webapp"
+        }
+        expression = "/archive/archive/dir";
+        nodes = (NodeList) xPath.compile(expression).evaluate(inputDoc, XPathConstants.NODESET);
+        Assert.assertEquals("Number of <archive/> element ==>", 3, nodes.getLength());
+        String sourceOnDisk1 = nodes.item(0).getAttributes().getNamedItem("sourceOnDisk").getNodeValue();
+        String sourceOnDisk2 = nodes.item(1).getAttributes().getNamedItem("sourceOnDisk").getNodeValue();
+        String sourceOnDisk3 = nodes.item(2).getAttributes().getNamedItem("sourceOnDisk").getNodeValue();
+
+        Assert.assertTrue("actual targetInArchive paths [" + sourceOnDisk1 + ","+ sourceOnDisk2 + ","
```
Comment (cherylking) [inline src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy:110]: message is misleading....targetInArchive should say sourceOnDisk.

File: src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy (hunk 14 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy:14]: This had 2 util methods which moved to DevTaskHelper. I have done an optimze import on this

**Review by cherylking — CHANGES_REQUESTED** [review]

File: src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy (hunk 14 above)
Comment (cherylking) [reply src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy:14]: but these new imports are not used by this file

File: src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy (hunk 15 above)
Comment (arunvenmany-ibm) [reply src/test/groovy/io/openliberty/tools/gradle/TestMultiModuleLooseEarWithPages.groovy]: corrected

File: src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy (hunk 14 above)
Comment (arunvenmany-ibm) [reply src/main/groovy/io/openliberty/tools/gradle/utils/LooseWarApplication.groovy]: corrected

**Review by cherylking — APPROVED** [review]
